### PR TITLE
Validator: Safer pending attestation records flushing

### PR DESCRIPTION
**Pull request by prestonvanloon**
**What type of PR is this?**

Other

**What does this PR do? Why is it needed?**

Better locking and flushing of this slice can help prevent dropped saves, in the event that there is already some batch save in progress.

**Which issues(s) does this PR fix?**

N/A

**Other notes for review**

### CODE REVIEW

File: validator/db/kv/attester_protection.go [hunk 1]
```diff
@@ -26,6 +27,37 @@ type AttestationRecord struct {
 	SigningRoot [32]byte
 }
 
+func NewPendingAttestationRecords() *PendingAttestationRecords {
```
Comment (rauljordan) [inline validator/db/kv/attester_protection.go:30]: This conflicts with PendingAttestation used as part of prysm's core package. Throughout our code base, they refer to attestations that are queued to enter the chain, but perhaps we could choose a different name

File: validator/db/kv/attester_protection.go [hunk 2]
```diff
@@ -258,13 +294,28 @@ func (s *Store) batchAttestationWrites(ctx context.Context) {
 // and resets the list of batched attestations for future writes.
 // This function notifies all subscribers for flushed attestations
 // of the result of the save operation.
-func (s *Store) flushAttestationRecords(ctx context.Context) {
+func (s *Store) flushAttestationRecords(ctx context.Context, records []*AttestationRecord) {
+	if s.batchedAttestationsFlushInProgress {
```
Comment (rauljordan) [inline validator/db/kv/attester_protection.go:298]: Atomic bool could be better here, we have it in the shared/ package

Comment (rauljordan) [reply validator/db/kv/attester_protection.go:298]: abool

Comment (prestonvanloon) [reply validator/db/kv/attester_protection.go]: Definitely, thanks!

File: validator/db/kv/attester_protection.go (hunk 1 above)
Comment (terencechain) [reply validator/db/kv/attester_protection.go:30]: It's actually an official eth2 object https://github.com/ethereum/eth2.0-specs/blob/dev/specs/phase0/beacon-chain.md#pendingattestation

Comment (prestonvanloon) [reply validator/db/kv/attester_protection.go:32]: Alright, I will rename

File: validator/db/kv/attester_protection.go [hunk 3]
```diff
@@ -258,13 +309,25 @@ func (s *Store) batchAttestationWrites(ctx context.Context) {
 // and resets the list of batched attestations for future writes.
 // This function notifies all subscribers for flushed attestations
 // of the result of the save operation.
-func (s *Store) flushAttestationRecords(ctx context.Context) {
+func (s *Store) flushAttestationRecords(ctx context.Context, records []*AttestationRecord) {
+	if s.batchedAttestationsFlushInProgress.IsSet() {
+		log.Error("Attempted to flush attestation records when already in progress")
```
Comment (rauljordan) [inline validator/db/kv/attester_protection.go:314]: Should this be at error level? maybe debug is appropriate. We should only have error logs when they are actionable to the user, and this is opaque from a user's perspective

File: validator/db/kv/attester_protection.go [hunk 4]
```diff
@@ -258,13 +309,25 @@ func (s *Store) batchAttestationWrites(ctx context.Context) {
...
-	err := s.saveAttestationRecords(ctx, s.batchedAttestations)
-	// If there was no error, we reset the batched attestations slice.
+	err := s.saveAttestationRecords(ctx, records)
+	// If there was any error, retry the records since the TX would have been reverted.
 	if err == nil {
 		log.WithField("duration", time.Since(start)).Debug("Successfully flushed batched attestations to DB")
-		s.batchedAttestations = make([]*AttestationRecord, 0, attestationBatchCapacity)
+	} else {
+		// This should never happen.
+		log.WithError(err).Error("Failed to batch save attestation records, retrying in queue")
```
Comment (rauljordan) [inline validator/db/kv/attester_protection.go:327]: Same ^

File: validator/db/kv/attester_protection.go (hunk 3 above)
Comment (prestonvanloon) [reply validator/db/kv/attester_protection.go:314]: This should never ever happen. If it is happening, then it's a development bug that was recently introduced. I will add a comment to reflect the significance

**Review by rauljordan — APPROVED** [review]

**Pull request merged**